pros/src/sensors/imu.rs

```diff
@@ -0,0 +1,414 @@
+            bail_on!(PROS_ERR, pros_sys::imu_set_euler(self.port, euler.into()));
+        }
+        Ok(())
+    }
+
+    /// Sets the current reading of the Inertial Sensor’s rotation to target value.
+    pub fn set_rotation(&self, rotation: f64) -> Result<(), InertialError> {
+        unsafe {
+            bail_on!(PROS_ERR, pros_sys::imu_set_rotation(self.port, rotation));
+        }
+        Ok(())
+    }
+
+    /// Sets the current reading of the Inertial Sensor’s heading to target value.
+    ///
+    /// Target will default to 360 if above 360 and default to 0 if below 0.
+    pub fn set_heading(&self, heading: f64) -> Result<(), InertialError> {
+        unsafe {
+            bail_on!(PROS_ERR, pros_sys::imu_set_heading(self.port, heading));
+        }
+        Ok(())
+    }
+
+    /// Sets the current reading of the Inertial Sensor’s pitch to target value.
+    ///
+    /// Will default to +/- 180 if target exceeds +/- 180.
+    pub fn set_pitch(&self, pitch: f64) -> Result<(), InertialError> {
+        unsafe {
+            bail_on!(PROS_ERR, pros_sys::imu_set_pitch(self.port, pitch));
+        }
+        Ok(())
+    }
+
+    /// Sets the current reading of the Inertial Sensor’s roll to target value
+    ///
+    /// Will default to +/- 180 if target exceeds +/- 180.
+    pub fn set_roll(&self, roll: f64) -> Result<(), InertialError> {
+        unsafe {
+            bail_on!(PROS_ERR, pros_sys::imu_set_roll(self.port, roll));
+        }
+        Ok(())
+    }
+
+    /// Sets the current reading of the Inertial Sensor’s yaw to target value.
+    ///
+    /// Will default to +/- 180 if target exceeds +/- 180.
+    pub fn set_yaw(&self, yaw: f64) -> Result<(), InertialError> {
+        unsafe {
+            bail_on!(PROS_ERR, pros_sys::imu_set_yaw(self.port, yaw));
+        }
+        Ok(())
+    }
+
+    pub fn set_data_rate(&self, data_rate: Duration) -> Result<(), InertialError> {
+        unsafe {
+            let rate_ms = if let Ok(rate) = u32::try_from(data_rate.as_millis()) {
+                rate
+            } else {
+                return Err(InertialError::InvalidDataRate);
+            };
+
+            bail_on!(PROS_ERR, pros_sys::imu_set_data_rate(self.port, rate_ms));
+        }
+        Ok(())
+    }
+}
+
+/// Standard quaternion consisting of a vector defining an axis of rotation
+/// and a rotation value about the axis.
+#[derive(Default, Debug, Clone, Copy, PartialEq)]
+pub struct Quaternion {
+    /// The x-component of the axis of rotation.
+    pub x: f64,
+
+    /// The y-component of the axis of rotation.
+    pub y: f64,
+
+    /// The z-component of the axis of rotation.
+    pub z: f64,
+
+    /// The magnitude of rotation about the axis.
+    pub w: f64,
+}
+
+impl TryFrom<pros_sys::quaternion_s_t> for Quaternion {
+    type Error = InertialError;
+
+    fn try_from(value: pros_sys::quaternion_s_t) -> Result<Quaternion, InertialError> {
+        Ok(Self {
+            x: bail_on!(PROS_ERR_F, value.x),
```

Comment:
I'm assuming there is no case where x wont be PROS_ERR_F, but the others will. Right?

Reply:
According to the PROS docs and what I can gather from the source code, yes.
- The struct is initialized as `QUATERNION_ERR_INIT` (populated with `PROS_ERR_F`) for all fields.
- ERRNO is only set if PROS can't claim the port (returned as PortError) in which case `QUATERNION_ERR_INIT` would be returned early.
- The struct is filled completely with data from `vexDeviceImuAttitudeGet`, which are assumed to be valid values by `pros::c` (no idea if the SDK even has error handling codes).
So there shouldn't be a case where the struct is somehow partially initialized with `PROS_ERR_F` in this case.